linker/shared/src/main/scala/org/scalajs/linker/backend/BasicLinkerBackend.scala

```diff
@@ -93,8 +107,141 @@ final class BasicLinkerBackend(config: LinkerBackendImpl.Config)
       }
     }
 
+    printedModuleSetCache.startRun()
+
     logger.timeFuture("BasicBackend: Write result") {
       writer.write(moduleSet)
+    }.map { result =>
+      printedModuleSetCache.cleanAfterRun()
+      printedModuleSetCache.logStats(logger)
+
+      result
+    }
+  }
+
+  private def extractLevelTrees(moduleBody: js.Tree): List[js.Tree] = {
+    moduleBody match {
+      case js.Block(stats) => stats
+      case js.Skip()       => Nil
+      case _               => moduleBody :: Nil
+    }
+  }
+}
+
+private object BasicLinkerBackend {
+  private final class PrintedModuleSetCache(withSourceMaps: Boolean) {
+    private val modules = new java.util.concurrent.ConcurrentHashMap[ModuleID, PrintedModuleCache]
+
+    private var totalTopLevelTrees = 0
+    private var recomputedTopLevelTrees = 0
+
+    def startRun(): Unit = {
+      totalTopLevelTrees = 0
+      recomputedTopLevelTrees = 0
+    }
+
+    def getModuleCache(moduleID: ModuleID): PrintedModuleCache = {
+      val result = modules.computeIfAbsent(moduleID, { _ =>
+        if (withSourceMaps) new PrintedModuleCacheWithSourceMaps(this)
+        else new PrintedModuleCache(this)
+      })
+
+      result.startRun()
+      result
+    }
+
+    def cleanAfterRun(): Unit = {
+      val iter = modules.entrySet().iterator()
+      while (iter.hasNext()) {
+        val moduleCache = iter.next().getValue()
+        if (moduleCache.cleanAfterRun()) {
+          totalTopLevelTrees += moduleCache.getTotalTopLevelTrees
+          recomputedTopLevelTrees += moduleCache.getRecomputedTopLevelTrees
+        } else {
+          iter.remove()
+        }
+      }
+    }
+
+    def logStats(logger: Logger): Unit = {
+      logger.debug(
+        s"BasicBackend: total top-level trees: $totalTopLevelTrees; re-computed: $recomputedTopLevelTrees")
+    }
+  }
+
+  private final class PrintedTree(val jsCode: String, val sourceMapFragment: SourceMapWriter.Fragment) {
```

Comment:
I can bring over the commit "Directly write bytes in the back-end instead of using a UTF-8 encoder." from #4817 over to this PR, if you prefer. I thought it would be easier to review it independently, but I can also bring it here.

Reply:
Actually, that commit addresses downstream concerns, but not what I meant here: It still prints the code to a string and then converts that to a byte array. I think we should do that processing in-stream.

Alternatively to keep it simple: Just add a TODO in the code here and we'll deal with it on the other PR.

Reply:
I see. OK, I added a TODO comment for now, then.